web/tests/visualization.spec.ts

```diff
@@ -0,0 +1,444 @@
+    }
+    return normalized(pos);
+}
+
+// Compare the coordinates with given precision. The compared coordinates
+// have to match on `precision` digits. For default we use 8 digits.
+function checkGraphPositions(saved, graph, precision) {
+    precision = precision || 8;
+    for (let i = 0; i < saved.length; ++i) {
+        expect(saved[i].x).toBeCloseTo(graph[i].x, precision);
+        expect(saved[i].y).toBeCloseTo(graph[i].y, precision);
+    }
+}
+
+
+test.beforeAll(async ({ browser }) => {
+    workspace = await Workspace.empty(await browser.newPage());
+    await workspace.addBox({ id: 'eg0', name: 'Create example graph', x: 100, y: 100 });
+    await workspace.addBox({
+        id: 'sg0',
+        name: 'Use base graph as segmentation',
+        x: 100, y: 200,
+        after: 'eg0',
+        params: { name: 'seg' },
+    });
+    await workspace.addBox({
+        id: 'vz0',
+        name: 'Graph visualization',
+        x: 100, y: 300,
+        after: 'sg0',
+    });
+    await (await workspace.openBoxEditor('vz0')).moveTo(800, 90);
+    await (await workspace.openStateView('vz0', 'visualization')).moveTo(270, 90)
+
+});
+
+test('sampled mode attribute visualizations', async () => {
+    const expectedEdges = [
+        { src: 0, dst: 1 },
+        { src: 1, dst: 0 },
+        { src: 2, dst: 0 },
+        { src: 2, dst: 1 },
+    ];
+    let savedPositions;
+    const DEFAULT = 'rgb(57, 188, 243)'; // Brand color.
+    const LOW = 'rgb(68, 1, 84)';
+    const HIGH = 'rgb(254, 232, 37)';
+
+    const visualization = workspace.getStateView('vz0', 'visualization').visualization;
+    const editor = workspace.getVisualizationEditor('vz0');
+    const name = editor.left.vertexAttribute('name');
+    const gender = editor.left.vertexAttribute('gender');
+    const income = editor.left.vertexAttribute('income');
+    const age = editor.left.vertexAttribute('age');
+    const location = editor.left.vertexAttribute('location');
+    const weight = editor.left.edgeAttribute('weight');
+    const comment = editor.left.edgeAttribute('comment');
+
+    // No attributes visualized.
+    let graph = await visualization.graphData();
+    await expect(workspace.page.locator('text=3 vertices')).toBeVisible();
+    await expect(graph.edges).toConcur(expectedEdges);
+    await expect(graph.edges).toConcur([
+        { color: '', label: '', width: '>2' },
+        { color: '', label: '', width: '>2' },
+        { color: '', label: '', width: '>2' },
+        { color: '', label: '', width: '>2' },
+    ]);
+    await expect(graph.vertices).toConcur([
+        { color: DEFAULT, icon: 'circle', label: '' },
+        { color: DEFAULT, icon: 'circle', label: '' },
+        { color: DEFAULT, icon: 'circle', label: '' },
+    ]);
+    savedPositions = positions(graph);
+
+    await name.visualizeAs('label');
+    await expect(workspace.page.locator('text=Label: name')).toBeVisible();
+    graph = await visualization.graphData();
```

Comment:
I'm afraid it's hard to make this reliable. You change a visualization setting. When do you fetch the graph data? When there is no loading animation? It could be that we're just early and the animation has not started yet! So we wait for the animation first? But it could be that we're late and the animation has already finished!

It's for exactly this reason that I replaced `expect(getHistogramValues())` with `expectHistogramValues()` in https://github.com/lynxkite/lynxkite/pull/356/commits/51c313972dcb912bd682d0c72b24ad380c6d1071. Instead of fetching some data and then making assertions about it, the better option with Playwright is to make assertions directly about the page. Those assertions wait to become true.

That said, this is a lot of tests here. It's a lot of work to fix them all like that. If they seem reliable enough as they are, I'm fine with leaving them as they are.